File: Java/mypack/MergeSort.java
//MergeSort algorithm
class MergeSort{
	int [] a = {1, -2, 3, -10, -29};
	int beg, end;

	void mergeSort(int[] a, int p, int r){
		if(p < r){
			int q = (p + r) / 2;
			mergeSort(a, p, q);
			mergeSort(a, q+1, r);
			merge(a, p, q, r);
		}
	}

	void merge(int[] a, int p, int q, int r){
		int m = q - p + 1, n = r - q;
		int [] a1,a2;
		a1 = new int[m+1];
		a2 = new int[n+1];

		for(int i=0;i<m;i++) a1[i] = a[p+i];
		for(int i=0;i<n;i++) a2[i] = a[q+i+1];
			a1[m] = a2[n] = Integer.MAX_VALUE;

		int t=0,j=0; 
		for(int k=p;k<=r;k++){
			if(a1[t] <= a2[j]){
				a[k] = a1[t];
				t++;
			}else{
				a[k] = a2[j];
				j++;
			}
		}
	}
	public static void main(String[] args){
		MergeSort ob = new MergeSort();
		System.out.print("Initial array: ");
		for(int x : ob.a) System.out.print(x + " ");
		ob.beg = 0;
		ob.end = ob.a.length - 1;
		//System.out.println("\n" + Integer.MAX_VALUE);
		ob.mergeSort(ob.a,ob.beg,ob.end);

		System.out.print("\nSorted array: ");
		for(int x : ob.a) System.out.print(x + " ");
	}

}

File: DSA lab/C/Mergesort.c
// Merge sort implementation
# include <stdio.h>
# define SIZE 7
void mergeSort (int [], int, int);
void merge (int [], int, int, int);

void mergeSort (int a[], int beg, int end) {
    if (beg < end) {
        int mid = (beg + end) / 2;
        mergeSort (a, beg, mid);
        mergeSort (a, mid+1, end);
        merge (a, beg, mid, end);
    }
}

void merge (int a[], int beg, int mid, int end) {
    int b[mid-beg+1];
    int c[end-mid];
    for (int i=0; i<=mid; i++)
        b[i] = a[beg+i];
    for (int i=0; i<end-mid; i++)
        c[i] = a[mid+i+1];
    
    int i=beg,j=mid+1,k=beg;
    while (i <= mid && j <= end) {
        if (b[i] <= c[j]) {
            a[k] = b[i];
            i++;
        }else {
            a[k] = c[j];
            j++;
        }
        k++;
    }
    if (i > mid) {
        for (int t=j; t<=end; t++) 
            a[k++] = c[j];
    }
    if (j > end) {
        for (int t=i; t<=mid; t++)
            a[k++] = b[t];
    }
}

void main () {
    int a[SIZE];
    printf("Enter %d elements in the array: ", SIZE);
    for (int i=0; i<SIZE; i++) {
        scanf ("%d", &a[i]);
    }
    printf("Given array: ");
    for (int i=0; i<SIZE; i++) {
        printf("%d ", a[i]);
    }
    mergeSort(a, 0, SIZE-1);

    printf("\nSorted array: ");
    for (int i=0; i<SIZE; i++) {
        printf("%d ", a[i]);
    }
}


